### DISCUSSION TITLE
Sharing the Selenium driver instance to another step on another file

### DISCUSSION THREAD

**Discussion by jkchua** [getgauge/gauge#2460] [General]
I have a current problem regarding **Selenium** with **Gauge** as the BDD framework using **NodeJS** as programming language. 

### **Context**

For instance, I have the following steps that I need to implement:

```
specification_one.spec

* Visit the site that I'm testing.
* Click on a button

specification_two.spec

* Visit the site that I'm testing.
* Click on a checkbox.
```

So in my example above, both specs have `Visit the site that I'm testing.` in order to visit the site on a different set of test cases. The following code for `Visit the site that I'm testing.` is as follows:

```
common_steps.js

let driver;

beforeSpec(function() {
    const edgeOptions = new Options();

    const preferences = {
        user_experience_metrics: {
            personalization_data_consent_enabled: true,
        },
    };
    edgeOptions.setUserPreferences(preferences);
    edgeOptions.excludeSwitches(["enable-logging"]);

    driver = await new Builder().forBrowser(Browser.EDGE).setEdgeOptions(edgeOptions).build();
}) 

step("Visit the site that I'm testing", async function () {
    await driver.get(process.env.BASE_URL);
    await driver.manage().setTimeouts({ implicit: parseInt(process.env.test_timeout) });
    await driver.manage().window().maximize();

    await driver.wait(until.titleIs("Title"), parseInt(process.env.test_timeout));
});
```

Now if I use this code in other specs, it will throw that the driver is undefined, which is what's happening now. 

### Questions

* How do I make the step `Visit the site that I'm testing.` be able to be used by other specifications with driver instance available to the next set of steps? My problem is I can't use the step to other specs because it's looking for a definition of the driver and I am not sure if hooks were the right way to do it (I ran out of ideas.)

**Comment by bentoboop** [comment]
Hello,

A workaround I have found for a similar issue was to store the object  in the gauge datastore in my BeforScenario hook, and map the variable in a before step in every JS file.

So when starting your driver you could use:

`gauge.dataStore.scenarioStore.put("current_driver", driver);
`
And then somewhere in every step implementation JS files:

```
let driver;

beforeStep(async () => {
    driver = gauge.dataStore.scenarioStore.get("current_driver");
});
```

It's problably not the most elegant way, and might consume unecessary ressources to run this BeforeStep every single step, but working with Playwright and multiple browser contexts in a single test, it was quite efficient. I needed a way to be able to call steps to switch from a browser to another from the spec file easily, while not having any step implemetations tied to a specific context, and it worked.

I hope it helps.

**Reply by jkchua** [reply]
Thank you for your kind response. So what I tried is to replicate the issue with Google as test site, and isolated the hooks in `hooks.js`, and tried to use `beforeSpec` and `specStore` in place of `scenarioStore` and it seems like it's not succeeding as expected. 

```
hooks.js

const { Builder, Browser, By, Key } = require("selenium-webdriver");
const assert = require("assert");
const { Options } = require("selenium-webdriver/edge");

let driver;

beforeSpec(async function () {
    const edgeOptions = new Options();
    const preferences = {
        user_experience_metrics: {
            personalization_data_consent_enabled: true,
        },
    };
    edgeOptions.setUserPreferences(preferences);
    edgeOptions.excludeSwitches(["enable-logging"]);
    driver = await new Builder().forBrowser(Browser.EDGE).setEdgeOptions(edgeOptions).build();

    await driver.manage().setTimeouts({ implicit: parseInt(10000) });
    await driver.manage().window().maximize();

    gauge.dataStore.specStore.put("current_driver", driver);
});

```

```
scenario_test_1.js

"use strict";

const { Builder, Browser, By, Key } = require("selenium-webdriver");
const assert = require("assert");
const { Options } = require("selenium-webdriver/edge");

let driver = gauge.dataStore.specStore.get("current_driver");

step("Visit Google.", async function () {
    await driver.get("https://www.google.com");
});

step("Enter <keyword> as the keyword.", async function (keyword) {
    let textbox = await driver.findElement(By.xpath("//*[@id='APjFqb']"));
    await textbox.sendKeys(keyword);
    await textbox.sendKeys(Key.ENTER);
});

step("Click the first result.", async function () {
    await driver.findElement(By.xpath("//*[@id='kp-wp-tab-overview']/div[3]/div/div/div/div/div/div[1]/div/div/span/a")).click();
});

```
```
scenario_test_2.js

"use strict";

const { Builder, Browser, By } = require("selenium-webdriver");
const assert = require("assert");

let driver = gauge.dataStore.specStore.get("current_driver");

step("Click the second result.", async function() {
    await driver.findElement(By.xpath("//*[@id='rso']/div[2]/div/div[2]/div/div/g-section-with-header/div[1]/div/div[1]/g-link/a")).click();
});
```

``` 
scenario_test_1.spec

# Scenario Test 1

This is an executable specification file. This file follows markdown syntax.
Every heading in this file denotes a scenario. Every bulleted point denotes a step.

## Test Scenario 1
tags: testing

* Visit Google.
* Enter "Gauge" as the keyword.
* Click the first result.
```

```
scenario_test_2.spec

# Scenario Test 2

This is an executable specification file. This file follows markdown syntax.
Every heading in this file denotes a scenario. Every bulleted point denotes a step.

## Test Scenario 2

* Visit Google.
* Enter "YouTube" as the keyword.
* Click the second result.
```

The error thrown is that it was null, which makes me think that the beforeSpec didn't initialize (but it somehow opened a browser with `data:,` as URL.

```
# Scenario Test 1
  ## Test Scenario 1     F
        Failed Step: Visit Google.
        Specification: specs\scenario_test_1.spec:9
        Error Message: {}
        Stacktrace:
        TypeError: Cannot read properties of null (reading 'get')
            at Object.<anonymous> (tests\scenario_test_1.js:10:18)
```

**Reply by bentoboop** [reply]
My understanding is that you cannot call the gauge functions outside a step:

`let driver = gauge.dataStore.specStore.get("current_driver");
`

has to be called either in a step definition, or in the BeforeStep example I gave.

Try : 

```
let driver;

step("Visit Google.", async function () {
    driver = gauge.dataStore.specStore.get("current_driver");
    await driver.get("https://www.google.com");
});
```

to see my point. The driver should be found at least for this step, but not the others.

**Reply by jkchua** [reply]
That made sense now! 😃

I tried your suggestion too, and it did work as expected, but I agree, it's not as elegant as it should be. I still used BeforeSpec to trigger Selenium initialization, and BeforeStep to fetch the driver. I'll have to monitor if it takes up resources as my example is just a Google search scenario.

You're a lifesaver by the way! Thank you very much, mate! 😊
